lib/rules/logical-assignment-operators.js

```diff
@@ -0,0 +1,465 @@
+ */
+function isUndefined(expression, scope) {
+    if (expression.type === "Identifier" && expression.name === "undefined") {
+        return astUtils.isReferenceToGlobalVariable(scope, expression);
+    }
+
+    return expression.type === "UnaryExpression" &&
+           expression.operator === "void" &&
+           expression.argument.type === "Literal" &&
+           expression.argument.value === 0;
+}
+
+/**
+ * Returns true iff the reference is either an identifier or member expression
+ * @param {ASTNode} expression Expression to check
+ * @returns {boolean} True for identifiers and member expressions
+ */
+function isReference(expression) {
+    return (expression.type === "Identifier" && expression.name !== "undefined") ||
+           expression.type === "MemberExpression";
+}
+
+/**
+ * Returns true iff the expression checks for nullish with loose equals.
+ * Examples: value == null, value == void 0
+ * @param {ASTNode} expression Test condition
+ * @param {import('eslint-scope').Scope} scope Scope of the expression
+ * @returns {boolean} True iff implicit nullish comparison
+ */
+function isImplicitNullishComparison(expression, scope) {
+    if (expression.type !== "BinaryExpression" || expression.operator !== "==") {
+        return false;
+    }
+
+    const reference = isReference(expression.left) ? "left" : "right";
+    const nullish = reference === "left" ? "right" : "left";
+
+    return isReference(expression[reference]) &&
+           (astUtils.isNullLiteral(expression[nullish]) || isUndefined(expression[nullish], scope));
+}
+
+/**
+ * Condition with two equal comparisons.
+ * @param {ASTNode} expression Condition
+ * @returns {boolean} True iff matches ? === ? || ? === ?
+ */
+function isDoubleComparison(expression) {
+    return expression.type === "LogicalExpression" &&
+           expression.operator === "||" &&
+           expression.left.type === "BinaryExpression" &&
+           expression.left.operator === "===" &&
+           expression.right.type === "BinaryExpression" &&
+           expression.right.operator === "===";
+}
+
+/**
+ * Returns true iff the expression checks for undefined and null.
+ * Example: value === null || value === undefined
+ * @param {ASTNode} expression Test condition
+ * @param {import('eslint-scope').Scope} scope Scope of the expression
+ * @returns {boolean} True iff explicit nullish comparison
+ */
+function isExplicitNullishComparison(expression, scope) {
+    if (!isDoubleComparison(expression)) {
+        return false;
+    }
+    const leftReference = isReference(expression.left.left) ? "left" : "right";
+    const leftNullish = leftReference === "left" ? "right" : "left";
+    const rightReference = isReference(expression.right.left) ? "left" : "right";
+    const rightNullish = rightReference === "left" ? "right" : "left";
+
+    return astUtils.isSameReference(expression.left[leftReference], expression.right[rightReference], true) &&
```

Comment:
Originally I added this as the rule `operator-assignments` also handles it this way.
This only affects cases when the computed access is a template literal or the access is different (eg. `a.b = a["b"] || c`).
I did some digging and the PR #13416 that introduced the utility was  to support optional chaining and the only comment is it is for backwards compability. So I think it as to not introduce new lint errors. As this is a new rule this would not matter, unless the rule should work like `operator-assignments` in this regard.
In any way I added test cases explicitly for this case and removed the argument .

Reply:
Yes, this extra param was added only for backwards compatibility at the time. We might want to remove it at some point. I agree - since this is a new rule, there are no backwards compatibility concerns so it can use ` astUtils.isSameReference` in the mode that will catch more errors (the default `disableStaticComputedKey = false` mode).